vllm/v1/core/kv_cache_metrics.py

```diff
@@ -0,0 +1,161 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""KV cache metrics tracking."""
+
+import random
+import threading
+import time
+from collections import deque
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from vllm.v1.core.kv_cache_utils import KVCacheBlock
+
+from vllm.v1.metrics.stats import BlockResidencyEvent
+
+
+class BlockMetricsState:
+    """Tracks lifecycle metrics for a single KV cache block."""
+
+    __slots__ = (
```

Comment:
Since we're not using this anywhere else in vLLM, this is probably a premature optimization. Interestingly, one of the few places it was used in the (now removed) v0 codebase was in #5602 for an equivalent object. However, with your sampling approach, I think the allocation overhead is even less likely to be an issue.